### PR TITLE
#3038 invoke content assist on trigger characters

### PR DESCRIPTION
**Pull request by evidolob**
### What does this PR do?
 invoke content assist on trigger characters
### What issues does this PR fix or reference?
#3038

Signed-off-by: Evgen Vidolob <evidolob@codenvy.com>

### CODE REVIEW

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/codeassist/LanguageServerCodeAssistProcessor.java [hunk 1]
```diff
@@ -100,6 +100,24 @@ public String getErrorMessage() {
         return lastErrorMessage;
     }
 
+    @Override
+    public char[] getTriggerCharacters() {
+        if(serverCapabilities.getCompletionProvider() !=null){
```
Comment (benoitf) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/codeassist/LanguageServerCodeAssistProcessor.java]: code formatting issue (spaces)

File: plugins/plugin-orion/che-plugin-orion-editor/src/main/java/org/eclipse/che/ide/editor/orion/client/OrionEditorInit.java [hunk 2]
```diff
@@ -28,11 +29,7 @@
 import org.eclipse.che.ide.api.editor.document.Document;
 import org.eclipse.che.ide.api.editor.document.DocumentHandle;
 import org.eclipse.che.ide.api.editor.editorconfig.TextEditorConfiguration;
-import org.eclipse.che.ide.api.editor.events.CompletionRequestEvent;
-import org.eclipse.che.ide.api.editor.events.CompletionRequestHandler;
-import org.eclipse.che.ide.api.editor.events.DocumentChangeEvent;
-import org.eclipse.che.ide.api.editor.events.TextChangeEvent;
-import org.eclipse.che.ide.api.editor.events.TextChangeHandler;
+import org.eclipse.che.ide.api.editor.events.*;
```
Comment (benoitf) [inline plugins/plugin-orion/che-plugin-orion-editor/src/main/java/org/eclipse/che/ide/editor/orion/client/OrionEditorInit.java]: we shouldn't use wildcard imports

File: plugins/plugin-orion/che-plugin-orion-editor/src/main/java/org/eclipse/che/ide/editor/orion/client/OrionEditorInit.java [hunk 3]
```diff
@@ -233,10 +254,28 @@ public void onCompletionRequest(final CompletionRequestEvent event) {
     }
 
     /**
+     * Add listener for document changes to auto invoke auto completion on trigger character
+     */
+    private void handleTriggerCharacters(final Set<Character> triggerCharacters, DocumentHandle documentHandle, final CodeAssistant codeAssistant) {
+        documentHandle.getDocEventBus().addHandler(DocumentChangeEvent.TYPE, new DocumentChangeHandler() {
+            @Override
+            public void onDocumentChange(DocumentChangeEvent event) {
+                if (editorPropertiesManager.getEditorProperties().get(EditorProperties.SHOW_CONTENT_ASSIST_AUTOMATICALLY.toString()).isBoolean().booleanValue()) {
```
Comment (vparfonov) [inline plugins/plugin-orion/che-plugin-orion-editor/src/main/java/org/eclipse/che/ide/editor/orion/client/OrionEditorInit.java]: Is possible NPE here?

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/codeassist/CodeAssistProcessor.java [hunk 4]
```diff
@@ -49,4 +50,10 @@
      * @return an error message or <code>null</code> if no error occurred
      */
     String getErrorMessage();
+
+    /**
+     * The characters that trigger completion automatically.
+     */
+    @Nullable
+    char[] getTriggerCharacters();
```
Comment (kaloyan-raev) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/codeassist/CodeAssistProcessor.java]: Please consider trigger character sequences of more than one char. See https://github.com/Microsoft/language-server-protocol/issues/138 for details. This means the return type should be `String[]`.

**Review by kaloyan-raev — CHANGES_REQUESTED** [review]

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/codeassist/CodeAssistProcessor.java (hunk 4 above)
Comment (evidolob) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/codeassist/CodeAssistProcessor.java]: OK, I will update pull request with this

**Review by kaloyan-raev — APPROVED** [review]

**Review by vparfonov — APPROVED** [review]

**Comment by TylerJewell** [comment]
@evidolob - please plan to merge asap.

**Comment by TylerJewell** [comment]
@vparfonov - what is the merge plan?

**Comment by vparfonov** [comment]
details https://github.com/eclipse/che/pull/4135

**Pull request closed without merging**